base/summarysize.jl

```diff
@@ -0,0 +1,148 @@
+immutable SummarySize
+    seen::ObjectIdDict
+    frontier_x::Vector{Any}
+    frontier_i::Vector{Int}
+    exclude::Any
+    chargeall::Any
+end
+
+
+"""
+    Base.summarysize(obj; exclude=Union{...}, chargeall=Union{...}) -> Int
+
+Compute the amount of memory used by all unique objects reachable from the argument.
+Keyword argument `exclude` specifies the types of objects to exclude from the traversal.
+Keyword argument `chargeall` specifies the types of objects to always charge the size of all of their fields, even if those fields would normally be excluded.
+"""
+function summarysize(obj::ANY; exclude::ANY = Union{DataType, TypeName}, chargeall::ANY = Union{TypeMapEntry, Method, Core.MethodInstance})
+    ss = SummarySize(ObjectIdDict(), Any[], Int[], exclude, chargeall)
+    size::Int = ss(obj)
+    while !isempty(ss.frontier_x)
+        # DFS heap traversal of everything without a specialization
+        # BFS heap traversal of anything with a specialization
+        x = ss.frontier_x[end]
+        i = ss.frontier_i[end]
+        val = nothing
+        if isa(x, SimpleVector)
+            nf = length(x)
+            if isassigned(x, i)
+                val = x[i]
+            end
+        elseif isa(x, Array)
+            nf = length(x)
+            if ccall(:jl_array_isassigned, Cint, (Any, UInt), x, i - 1) != 0
+                val = x[i]
+            end
+        else
+            nf = nfields(x)
+            ft = typeof(x).types
+            if !isbits(ft[i]) && isdefined(x, i)
+                val = getfield(x, i)
+            end
+        end
+        if nfields(x) > i
+            ss.frontier_i[end] = i + 1
+        else
+            pop!(ss.frontier_x)
+            pop!(ss.frontier_i)
+        end
+        if val !== nothing && !isa(val, Module) && (!isa(val, ss.exclude) || isa(x, ss.chargeall))
+            size += ss(val)::Int
+        end
+    end
+    return size
+end
+
+(ss::SummarySize)(obj::ANY) = _summarysize(ss, obj)
+# define the general case separately to make sure it is not specialized for every type
+@noinline function _summarysize(ss::SummarySize, obj::ANY)
+    key = pointer_from_objref(obj)
+    haskey(ss.seen, key) ? (return 0) : (ss.seen[key] = true)
+    if nfields(obj) > 0
+        push!(ss.frontier_x, obj)
+        push!(ss.frontier_i, 1)
+    end
+    if isa(obj, UnionAll)
+        # black-list of items that don't have a Core.sizeof
+        return 2 * sizeof(Int)
```

Comment:
the old computation for UnionAll also added `summarysize(obj.body, seen, excl) + summarysize(obj.var, seen, excl)`, is this new version undercounting?